### PR TITLE
Add support NXP i.MX8ULP

### PR DESCRIPTION
**Pull request by TangleZ**
Will send more fixes in order for the support to be complete, the reset will be sent later.
The rest patches are doing the review for this on Github with SOF people.

### CODE REVIEW

**Comment by dbaluta** [comment]
Please fix checkpatch.pl errors. https://github.com/thesofproject/sof/pull/4550/checks?check_run_id=3167561201

Of course, we will NOT edit OS files coming from Cadence like core-isa.h. But there are a lot of NXP code which we should fix.

**Comment by marc-hb** [comment]
> Of course, we will NOT edit OS files coming from Cadence like core-isa.h. But there are a lot of NXP code which we should fix.

Not a big deal but in general I prefer separating copies versus new code in different commits. This makes code style checks like checkpatch.pl easier but not just: it also leverages git to record what code is original versus just a copy which helps identifying out-of-date copies later, makes code reviews easier, etc. BTW commit message(s) for copied code should include some version or tag referring to what this is a copy of.

The ideal version control to avoid copying code is something like https://docs.zephyrproject.org/latest/guides/modules.html but that can be overkill for just a few files.

**Comment by dbaluta** [comment]
@marc-hb  true, this is ideal situation. Anyhow,  we followed initialed support enabling for other platforms. If you really think this would make a difference we can split the patches.

File: src/platform/imx8ulp/include/platform/lib/dma.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * Author: Peng Zhang <peng.zhang_8@nxp.com>
+ */
+
+#ifdef __SOF_LIB_DMA_H__
+
+#ifndef __PLATFORM_LIB_DMA_H__
+#define __PLATFORM_LIB_DMA_H__
+
+#define PLATFORM_NUM_DMACS	3
+
```
Comment (dbaluta) [inline src/platform/imx8ulp/include/platform/lib/dma.h:14]: Is this really 3? We have EDMA and Dummy DMA. Am I missing something?

File: src/platform/imx8ulp/lib/memory.c [hunk 2]
```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright 2021 NXP
+//
+// Author: Daniel Baluta <daniel.baluta@nxp.com>
+
```
Comment (dbaluta) [inline src/platform/imx8ulp/lib/memory.c:6]: Fix author here.

File: src/arch/xtensa/CMakeLists.txt [hunk 3]
```diff
@@ -29,6 +29,8 @@ elseif(CONFIG_IMX8X)
 	set(platform_folder imx8)
 elseif(CONFIG_IMX8M)
 	set(platform_folder imx8m)
+elseif(CONFIG_IMX8ULP)
+	set(platform_folder imx8ulp)
 endif()
```
Comment (dbaluta) [inline src/arch/xtensa/CMakeLists.txt:34]: This should be in previous patch.

File: src/platform/CMakeLists.txt [hunk 4]
```diff
@@ -23,6 +23,8 @@ elseif(CONFIG_IMX8 OR CONFIG_IMX8X)
 	add_subdirectory(imx8)
 elseif(CONFIG_IMX8M)
 	add_subdirectory(imx8m)
+elseif(CONFIG_IMX8ULP)
+	add_subdirectory(imx8ulp)
 endif()
```
Comment (dbaluta) [inline src/platform/CMakeLists.txt:28]: same this should be in previous patch.

File: src/platform/Kconfig [hunk 5]
```diff
@@ -214,6 +214,23 @@ config IMX8M
...
+	select WAITI_DELAY
+	select IMX
+	select IMX_EDMA
+	select IMX_INTERRUPT_GENERIC
+	select SCHEDULE_DMA_MULTI_CHANNEL
+	help
+	 Select if your target platform is imx8ulp-compatible.
+	 imx.8ulp support dsp.
+
 endchoice
```
Comment (dbaluta) [inline src/platform/Kconfig:234]: ditto.

File: src/platform/Kconfig [hunk 6]
```diff
@@ -370,6 +387,7 @@ config RIMAGE_SIGNING_SCHEMA
 	default "imx8" if IMX8
 	default "imx8x" if IMX8X
 	default "imx8m" if IMX8M
+	default "imx8ulp" if IMX8ULP
```
Comment (dbaluta) [inline src/platform/Kconfig:390]: ditto.

File: scripts/xtensa-build-all.sh [hunk 7]
```diff
@@ -6,7 +6,7 @@
 set -e
 
 SUPPORTED_PLATFORMS=(byt cht bdw hsw apl skl kbl cnl sue icl jsl \
-                    imx8 imx8x imx8m tgl tgl-h)
```
Comment (dbaluta) [inline scripts/xtensa-build-all.sh:9]: This patch should only contain the change for `scripts/xtensa-build-all.sh .

CMakeLists and Kconfig should be updated in previous patch together with adding ``*c`` and ``*h`` files.

File: src/platform/imx8ulp/lib/memory.c (hunk 2 above)
Comment (TangleZ) [reply src/platform/imx8ulp/lib/memory.c:6]: But this file I haven't modified it, so does it also needs to change the author?

File: scripts/xtensa-build-all.sh (hunk 7 above)
Comment (TangleZ) [reply scripts/xtensa-build-all.sh:9]: OK. will modify this.

File: src/platform/imx8ulp/include/platform/lib/dma.h (hunk 1 above)
Comment (TangleZ) [reply src/platform/imx8ulp/include/platform/lib/dma.h:14]: It should be 2. Will update.

File: src/platform/imx8ulp/lib/memory.c (hunk 2 above)
Comment (dbaluta) [reply src/platform/imx8ulp/lib/memory.c:6]: Sure, but you added a new file. So at least add yourself to the authors.

File: src/drivers/imx/Kconfig [hunk 8]
```diff
@@ -13,3 +13,10 @@ config IMX_INTERRUPT_IRQSTEER
 	help
 	  This has to be selected for i.MX NXP platform that have irqsteer.
 	  It enables NXP platforms-specific features.
+
+config IMX_INTERRUPT_GENERIC
+	bool
+	default n
+	help
+	  This has to be selected for i.MX NXP platform that haven't irqsteer.
```
Comment (iuliana-prodan) [inline src/drivers/imx/Kconfig:21]: s/haven't/don't have

File: src/platform/imx8ulp/include/platform/drivers/idc.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright 2019 NXP
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/drivers/idc.h:3]: Update copyright year and author.

File: src/drivers/imx/interrupt-generic.c [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#include <stddef.h>
+#include <stdint.h>
+
+/* fa00558c-d653-4851-a03a-b21f125a9524 */
+DECLARE_SOF_UUID("generic-irq-imx", generic_irq_imx_uuid, 0xfa00558c, 0xd653, 0x4851,
+		 0xa0, 0x3a, 0xb2, 0x1f, 0x12, 0x5a, 0x95, 0x24);
+
+DECLARE_TR_CTX(noirq_i_tr, SOF_UUID(generic_irq_imx_uuid), LOG_LEVEL_INFO);
+
+int irqstr_get_sof_int(int irqstr_int)
```
Comment (iuliana-prodan) [inline src/drivers/imx/interrupt-generic.c:27]: Since this is the generic interrupt, I believe this should be 
```
int platform_get_interrupt(int irq)
{
	return irq;
}
```
Don't mention irqstr (from irq steer) anymore. :)

File: src/platform/imx8ulp/include/platform/drivers/interrupt.h [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+#define IRQ_SAI6_NUM         24
+#define IRQ_SAI7_NUM         25
+
+#define PLATFORM_IRQ_HW_NUM	XCHAL_NUM_INTERRUPTS
+#define PLATFORM_IRQ_CHILDREN	0 /* Each cascaded struct covers 64 IRQs */
+/* IMX: Covered steer IRQs are modulo-64 aligned. */
+#define PLATFORM_IRQ_FIRST_CHILD  0
+
+/* no irqstr in 8ulp */
+int irqstr_get_sof_int(int irqstr_int);
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/drivers/interrupt.h:33]: Don't mention irqstr (from irq steer) anymore.
Rename this function, see comment from [here](https://github.com/thesofproject/sof/pull/4550/commits/a89c6034138cfaf7ffc9866ef41a7ba272bbec02#diff-8800b4fb804080ade4c5978e9462553c70928737bb61b1bc9a9cd2f39614d22dR27).

File: src/platform/imx8ulp/include/platform/lib/clk.h [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <stdint.h>
+
+struct sof;
+
+#define CLK_CPU(x)	(x)
+
+#define CPU_DEFAULT_IDX		0
+
+#define CLK_DEFAULT_CPU_HZ	528000000
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/clk.h:21]: Align with below value

File: src/platform/imx8ulp/include/platform/lib/cpu.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2021 NXP Corporation. All rights reserved.
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/cpu.h:3]: `Copyright(c) 2021 NXP` I believe is enough.
Keep it  as in the other files

File: src/platform/imx8ulp/include/platform/lib/dai.h [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Author: Zhang Peng <peng.zhang_8@nxp.com>
+ */
+
+#ifdef __SOF_LIB_DAI_H__
+
+#ifndef __PLATFORM_LIB_DAI_H__
+#define __PLATFORM_LIB_DAI_H__
+
+#define DMAMUX2_SAI5_RX_NUM 69
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/dai.h:13]: Can you please add a comment for what are these for?
I see we we don't have any DAIs in the `dai.h` files for imx8mp or imx8.

Where and for what are these used?

File: src/platform/imx8ulp/include/platform/lib/dma.h [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+/* max number of supported DMA channels */
+#define PLATFORM_MAX_DMA_CHAN	32
+
+#define DMA_ID_EDMA2	0
+#define DMA_ID_HOST	1
+
+#define dma_chan_irq(dma, chan) \
+	(((int *)dma->plat_data.drv_plat_data)[chan])
+#define dma_chan_irq_name(dma, chan) NULL
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/dma.h:23]: Is this supposed to be NULL here? Maybe` dma_irq_name(dma)`.
Why does it has to be null?

File: src/platform/imx8ulp/include/platform/lib/mailbox.h [hunk 16]
```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright 2021 NXP
+ *
+ * Author: Peng Zhang
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/mailbox.h:5]: Be consistent with the author name and mail address.
Keep it as in the other files.

File: src/platform/imx8ulp/include/platform/lib/mailbox.h [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+ * +---------------------+----------------+-----------------------------------+
+ */
+
+#define MAILBOX_DSPBOX_SIZE		SRAM_OUTBOX_SIZE
+#define MAILBOX_DSPBOX_BASE		SRAM_OUTBOX_BASE
+#define MAILBOX_DSPBOX_OFFSET		SRAM_OUTBOX_OFFSET
+
+#define MAILBOX_HOSTBOX_SIZE		SRAM_INBOX_SIZE
+#define MAILBOX_HOSTBOX_BASE		SRAM_INBOX_BASE
+#define MAILBOX_HOSTBOX_OFFSET		SRAM_INBOX_OFFSET
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/mailbox.h:38]: Please try to align.

File: src/platform/imx8ulp/include/platform/lib/mailbox.h [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+#define MAILBOX_DEBUG_SIZE		SRAM_DEBUG_SIZE
+#define MAILBOX_DEBUG_BASE		SRAM_DEBUG_BASE
+#define MAILBOX_DEBUG_OFFSET		SRAM_DEBUG_OFFSET
+
+#define MAILBOX_TRACE_SIZE		SRAM_TRACE_SIZE
+#define MAILBOX_TRACE_BASE		SRAM_TRACE_BASE
+#define MAILBOX_TRACE_OFFSET		SRAM_TRACE_OFFSET
+
+#define MAILBOX_EXCEPTION_SIZE		SRAM_EXCEPT_SIZE
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/mailbox.h:48]: Here too

File: src/platform/imx8ulp/include/platform/lib/memory.h [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+ * | SOF_STACK_BASE      |                |                                   |
+ * +---------------------+----------------+-----------------------------------+
+ */
+
+/* Mailbox configuration */
+#define SRAM_OUTBOX_BASE	SDRAM1_BASE
+#define SRAM_OUTBOX_SIZE	0x1000
+#define SRAM_OUTBOX_OFFSET	0
+
+#define SRAM_INBOX_BASE		(SRAM_OUTBOX_BASE + SRAM_OUTBOX_SIZE)
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/memory.h:86]: Can you please try to align the values?

File: src/platform/imx8ulp/include/platform/lib/memory.h [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+
+/* Stack configuration */
+#define SOF_STACK_SIZE		0x1000
+#define SOF_STACK_TOTAL_SIZE	SOF_STACK_SIZE
+#define SOF_STACK_BASE		(SDRAM1_BASE + SDRAM1_SIZE)
+#define SOF_STACK_END		(SOF_STACK_BASE - SOF_STACK_TOTAL_SIZE)
+
+/* Vector and literal sizes - not in core-isa.h */
+#define SOF_MEM_VECT_LIT_SIZE		0x4
+#define SOF_MEM_VECT_TEXT_SIZE		0x1c
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/memory.h:167]: Where did you take these values from?

File: src/platform/imx8ulp/include/platform/lib/memory.h [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+#define SOF_MEM_VECT_SIZE		(SOF_MEM_VECT_TEXT_SIZE + SOF_MEM_VECT_LIT_SIZE)
+
+#define SOF_MEM_RESET_TEXT_SIZE		0x2e0
+#define SOF_MEM_RESET_LIT_SIZE		0x120
+#define SOF_MEM_VECBASE_LIT_SIZE	0x178
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/memory.h:172]: align

File: src/platform/imx8ulp/include/platform/lib/pm_runtime.h [hunk 22]
```diff
@@ -0,0 +1,61 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright 2019 NXP
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/platform/lib/pm_runtime.h:3]: It's a new file - update copyright and author.

File: src/platform/imx8ulp/lib/clk.c [hunk 23]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 NXP. All rights reserved.
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/lib/clk.c:3]: just `Copyright(c) 2021 NXP`

**Review by iuliana-prodan — CHANGES_REQUESTED** [review]

File: src/include/sof/drivers/mu.h [hunk 24]
```diff
@@ -13,27 +13,61 @@
...
+enum imx_mu_xsr {
+	IMX_MU_SR = 0x0c,
+	IMX_MU_GSR = 0x118,
+	IMX_MU_TSR = 0x124,
+	IMX_MU_RSR = 0x12c,
+};
+
+#ifdef CONFIG_IMX8ULP
+/* Transmit Register */
+#define IMX_MU_xTRn(x)         (0x200 + 4 * (x))
```
[the remark this answers is not in the document]
Comment (TangleZ) [reply src/include/sof/drivers/mu.h:43]: We do not use the transmit or receive register, but I suggest that to keep it in case we will need it in the future.

File: src/platform/imx8ulp/include/platform/drivers/interrupt.h (hunk 11 above)
Comment (TangleZ) [reply src/platform/imx8ulp/include/platform/drivers/interrupt.h:33]: Afraid we can't. The function irqstr_get_sof_int is called in esai.c, if we rename this function in imx8ulp platform, there will have a compiler error.

File: src/platform/imx8ulp/include/platform/lib/dai.h (hunk 14 above)
Comment (TangleZ) [reply src/platform/imx8ulp/include/platform/lib/dai.h:13]: It's in imx8ulp platform, only used in imx8ULP platform.

File: src/platform/imx8ulp/include/platform/lib/dma.h (hunk 15 above)
Comment (TangleZ) [reply src/platform/imx8ulp/include/platform/lib/dma.h:23]: Will change it.

File: src/platform/imx8ulp/include/platform/lib/memory.h (hunk 20 above)
Comment (TangleZ) [reply src/platform/imx8ulp/include/platform/lib/memory.h:167]: This part is just a copy from the imx8m platform.

File: src/drivers/imx/interrupt-generic.c (hunk 10 above)
Comment (dbaluta) [reply src/drivers/imx/interrupt-generic.c:27]: @TangleZ @iuliana-prodan I think Peng version is OK but we need some explanation

/* TODO: Fix this. Now this is required because ESAI driver assumes all i.MX8 platforms use irqsteer */

File: src/platform/imx8ulp/include/platform/drivers/idc.h (hunk 9 above)
Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/drivers/idc.h:3]: @TangleZ even if you copy a file from another platform please add your name to the authors lists. You can keep mine if you want and also add a comment something like  based on ``` src/platform/imx8m/include/platform/drivers/idc.h```

File: src/platform/imx8ulp/include/platform/drivers/interrupt.h (hunk 11 above)
Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/drivers/interrupt.h:33]: @TangleZ just add a comment like this: ``FIXME: no irqstr in 8ulp but we must implement this function because it is requred by ESAI */

File: src/platform/imx8ulp/include/platform/lib/cpu.h (hunk 13 above)
Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/lib/cpu.h:3]: @TangleZ 

Current guidelines from NXP is to use:

```
  * Copyright 2021 NXP
  ```

File: src/platform/imx8ulp/include/platform/lib/dai.h (hunk 14 above)
Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/lib/dai.h:13]: @TangleZ where are these used? What exactly are these? 
I would expect a small explanation.

Please add a comment something like this **WARN: I don't know the hardware**
```
/* DMAMUX_ settings for mapping DMA channels to DAIs */
```

Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/lib/dai.h:13]: On a second thought, please move DMAMUX stuff in a separate patch and explain why are they needed.

File: src/platform/imx8ulp/include/platform/lib/mailbox.h (hunk 17 above)
Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/lib/mailbox.h:38]: Align is OK as far as i can tell. The github GUI makes it look wrong.

File: src/platform/imx8ulp/include/platform/lib/memory.h (hunk 19 above)
Comment (dbaluta) [reply src/platform/imx8ulp/include/platform/lib/memory.h:86]: Looks good if you fetch the PR and look at it with an editor.

File: src/platform/imx8ulp/include/platform/trace/trace.h [hunk 25]
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2019 Intel Corporation. All rights reserved.
+ *
+ * Author: Tomasz Lauda <tomasz.lauda@linux.intel.com>
+ */
```
Comment (dbaluta) [inline src/platform/imx8ulp/include/platform/trace/trace.h:6]: same here. If you add a new file (even if it is copy pasted) please add your name and NXP copyright.
You can also mentioned the initial file you based your copy on .

File: src/platform/imx8ulp/lib/dai.c [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+		.fifo[SOF_IPC_STREAM_CAPTURE] = {
+			.offset		= SAI_5_BASE + REG_SAI_RDR2,
+			.depth		= 16,  /* in 4 bytes words */
+			.watermark      = 8,
+			.handshake	= EDMA_HANDSHAKE(EDMA2_8ULP_CHAN1_IRQ,
+							 EDMA2_8ULP_CHAN1),
+		},
+		.dmamux_rx_num = DMAMUX2_SAI5_RX_NUM,
+		.dmamux_tx_num = DMAMUX2_SAI5_TX_NUM,
+	},
```
Comment (dbaluta) [inline src/platform/imx8ulp/lib/dai.c:43]: Please move the changes with dmamux_rx_num in a separate patch together with the changes that add DMAMUX2_SAI5_TX_NUM. 

This is important to be in a separate patch because it touches core structures. In this huge PR it can be easily missed by Intel people.

They will be most likely to be interested in core changes and they will go easy on IMX changes.

Comment (dbaluta) [reply src/platform/imx8ulp/lib/dai.c:43]: Later edit: not sure if this even compiles. We need to modify `struct dai` right?

Comment (TangleZ) [reply src/platform/imx8ulp/lib/dai.c:43]: Yes. Need to modify dai first.

Comment (dbaluta) [reply src/platform/imx8ulp/lib/dai.c:43]: @TangleZ as discussed internally, please do not bother adding this stuff in the SOF core (dmamux, srcid). This is can be easily sorted out in the EDMA driver with an ifdef IMX8ULP. 
I think you were on a right path by trying to make this generic but this touches kernel core and if we want to have something with quality and working we can make this compromise now.

File: src/platform/imx8ulp/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+add_subdirectory(lib)
```
Comment (dbaluta) [inline src/platform/imx8ulp/CMakeLists.txt:3]: On a second thought please split this into two patches:
* one that adds Xtensa specific headers
* the rest

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
Looks good Peng. checkpatch.pl errors are mostly becasue we addded the Cadence header files.

**Review by paulstelian97 — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Review by iuliana-prodan — APPROVED** [review]

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@lgirdwood 

On-device Test Failed:APL_UP2_NOCODEC_ZEPHYR: kernel boot 

this looks unrelated.

Can you please merge this?

**Comment by dbaluta** [comment]
SOFCI TEST
